docs/language/informal/toplevel-inference.md

```diff
@@ -0,0 +1,311 @@
+
+
+For each candidate, inference for all of its dependencies must be performed
+before inference for the candidate is performed.  If this cannot be done
+(because of a cycle) it is a static error.  Note that it is always possible for
+the programmer to break a cycle and hence eliminate the error by adding a type
+annotation.
+
+
+### Candidate inference
+
+#### Top level variables
+
+The inferred type of a top level variable is the type inferred from its
+initializer.
+
+#### Static class variables
+
+The inferred type of a static class variable is the type inferred from its
+initializer.
+
+#### Instance field, getter, and setter inference
+
+
+The inferred type of a getter, setter, or field is as follows.  Note that I say
+that a setter overrides a getter if there is a getter of the same name in some
+superclass or interface, and similarly for setters overriding getters, fields,
+etc.
+
+
+A getter, setter or field which overrides/implements only a getter is inferred
+to have the type taken from the overridden getter result type.
+
+
+A getter, setter or field which overrides/implements only a setter is inferred
+to have the type taken from the overridden setter parameter.
+
+
+A getter which overrides/implements both a setter and a getter is inferred to
+have the type taken from the overridden getter result type.
+
+
+A setter which overrides/implements both a setter and a getter is inferred to
+have the type taken from the overridden setter parameter type.
+
+
+A field which overrides/implements both a setter and a getter is inferred to
+have the type taken from the overridden setter parameter type if this type is
+the same as the return type of the overridden getter (if the types are not the
+same then inference fails with an error).
+
+
+Note that overriding a field is addressed via the implicit induced getter/setter
+pair (or just getter in the case of a final field).
+
+
+A getter or setter with no annotated type that does not override anything is
+treated as if it were annotated with dynamic.
+
+
+A field with no annotated type that does not override anything has the type
+inferred from its initializer.
+
+
+### Initializer inference
+
+
+Initializer inference works by defining a subset of the expressions in the
+language for which we can predictably do inference in a way that is not
+sensitive to ordering outside of the static dependencies that are syntactically
+apparent in the expression.  In particular, field inference cannot change the
+outcome of initializer inference.
+
+
+A meta-goal is to keep this inference relatively lightweight and efficient to
+implement.
+
+
+In all cases, if initializer inference fails, it is an error and produces no
+result.  Implementations are free to treat this as a result of dynamic after
+emitting the error for the purposes of emitting subsequent error messages, but
+the behavior of programs for which initializer inference fails is unspecified.
+
+#### Immediately evident expressions
```

Comment:
I'd prefer 'Expressions with an immediately evident type', such that the reader gets the most informative phrase as the canonical one, up front, and then the abbreviated forms can be used more freely later on.